velox/functions/sparksql/fuzzer/SparkQueryRunner.cpp

```diff
@@ -0,0 +1,388 @@
+
+  grpc::ClientContext clientContext;
+  auto reader = stub_->ExecutePlan(&clientContext, *request);
+  ExecutePlanResponse response;
+  std::vector<RowVectorPtr> results;
+  while (reader->Read(&response)) {
+    if (response.session_id() == kSessionId && response.has_arrow_batch()) {
+      const std::string& data = response.arrow_batch().data();
+      const auto batchResults = readArrowData(data);
+      results.insert(results.end(), batchResults.begin(), batchResults.end());
+    }
+  }
+  return results;
+}
+
+std::vector<RowVectorPtr> SparkQueryRunner::readArrowData(
+    const std::string& data) {
+  auto buffer = std::make_shared<arrow::Buffer>(data);
+  auto bufferReader = std::make_shared<arrow::io::BufferReader>(buffer);
+  arrow::Result<std::shared_ptr<arrow::ipc::RecordBatchReader>> result =
+      arrow::ipc::RecordBatchStreamReader::Open(bufferReader);
+  VELOX_CHECK(
+      result.ok(),
+      "Failed to open RecordBatchReader: {}.",
+      result.status().ToString());
+
+  std::shared_ptr<arrow::ipc::RecordBatchReader> reader = result.ValueUnsafe();
+  std::shared_ptr<arrow::Schema> schema = reader->schema();
+
+  ArrowSchema arrowSchema;
+  auto status = arrow::ExportSchema(*schema, &arrowSchema);
+  VELOX_CHECK(status.ok(), "Failed to export schema: {}.", status.ToString());
+
+  arrow::Result<std::shared_ptr<arrow::RecordBatch>> batchResult;
+  std::vector<RowVectorPtr> results;
+  while ((batchResult = reader->Next()).ok() && batchResult.ValueUnsafe()) {
+    std::shared_ptr<arrow::RecordBatch> batch = batchResult.ValueUnsafe();
+    ArrowArray arrowArray;
+    status = ExportRecordBatch(*batch, &arrowArray);
+    VELOX_CHECK(status.ok(), "Failed to export array: {}.", status.ToString());
+    auto rv = std::dynamic_pointer_cast<RowVector>(
+        importFromArrowAsOwner(arrowSchema, arrowArray, pool()));
+    auto copy =
+        BaseVector::create<RowVector>(rv->type(), rv->size(), copyPool_.get());
+    copy->copy(rv.get(), 0, 0, rv->size());
+    results.push_back(copy);
+  }
+  VELOX_CHECK(
+      batchResult.ok(),
+      "Failed to read batch: {}.",
+      batchResult.status().ToString());
+  return results;
+}
+
+bool SparkQueryRunner::supported(
+    const std::shared_ptr<const core::AggregationNode>& aggregationNode) {
+  for (const auto& aggregate : aggregationNode->aggregates()) {
+    // Spark aggregation does not support sort keys.
+    if (!aggregate.sortingKeys.empty()) {
+      return false;
+    }
+    const auto functionName = aggregate.call->name();
+    if (functionName == "max_by" || functionName == "min_by") {
+      VELOX_CHECK_EQ(aggregate.rawInputTypes.size(), 2);
+      if (aggregate.rawInputTypes[1]->kind() == TypeKind::MAP) {
+        // Spark's `max_by` and `min_by` do not support ordering on map type.
+        return false;
+      }
+    }
+    if (aggregate.distinct) {
+      for (const auto& inputType : aggregate.rawInputTypes) {
+        // In Spark, grouping keys cannot be map type.
```

Comment:
Also, wondering if it is necessary to have this check here. We can instead just assume that plan is "valid" and let Spark fail if not. The Fuzzer needs to be configured properly to avoid generating "invalid" plans.